### PR TITLE
DX-1019: Read Your Writes Support

### PR DESCRIPTION
**Pull request by fahreddinozcan**

### CODE REVIEW

File: pkg/http.ts [hunk 1]
```diff
@@ -15,10 +15,13 @@ export type UpstashRequest = {
    * Request body will be serialized to json
    */
   body?: unknown;
+
```
Comment (ogzhanolguncu) [inline pkg/http.ts:18]: Maybe we can put a comment explains what this does.

File: pkg/redis.ts [hunk 2]
```diff
@@ -208,6 +209,7 @@ export class Redis {
     this.opts = opts;
     this.enableTelemetry = opts?.enableTelemetry ?? true;
     this.enableAutoPipelining = opts?.enableAutoPipelining ?? false;
+    this.readYourWrites = opts?.readYourWrites ?? true;
```
Comment (ogzhanolguncu) [inline pkg/redis.ts:212]: Maybe we should make this opt-in?

Comment (fahreddinozcan) [reply pkg/redis.ts:212]: I've talked with the core team, and they said there's literally no downside other than a small case. Most users wouldn't need to opt-out, and if they do it's possible.

**Review by CahidArda — APPROVED** [review]

File: pkg/http.ts [hunk 3]
```diff
@@ -202,6 +223,11 @@ export class HttpClient implements Requester {
       backend: this.options?.backend,
     };
 
+    if (this.readYourWrites) {
+      const newHeader = this.upstashSyncToken;
+      this.headers["upstash-sync-token"] = newHeader;
+    }
+
```
Comment (ogzhanolguncu) [inline pkg/http.ts:230]: Can we omit this and do header init in the if below? This a bit hard to follow. Felt unintentional

**Review by ogzhanolguncu — COMMENTED** [review]
Make sure to run `bun fmt ` before commit. Sorry for inconvenience we'll fix that stupid formatting issue sson.

File: pkg/http.ts (hunk 3 above)
Comment (fahreddinozcan) [reply pkg/http.ts:230]: There are two steps here,
* For each request, we set the header from the header of the previous requests' response, which is set as `upstashSyncToken` after the fetch request.
* After the fetch request, we get the response header and save it for the next object.

Since we don't have a global header object, we can't update the header in the second if, since we'd have already sent the request. And we also don't have a global header config, and we arrange that per request.

Comment (ogzhanolguncu) [reply pkg/http.ts:230]: Maybe add a small comment documenting this.

File: pkg/http.ts [hunk 4]
```diff
@@ -15,10 +15,21 @@ export type UpstashRequest = {
...
    */
   body?: unknown;
+
+  upstashSyncToken?: string;
 };
 export type UpstashResponse<TResult> = { result?: TResult; error?: string };
 
 export interface Requester {
+  /**
+  * When this flag is not disabled, new commands of this client expects the previous write commands to be finalized before executing.
```
Comment (mdogan) [inline pkg/http.ts:25]: When this flag is enabled, any subsequent commands issued by this client are guaranteed to observe the effects of all earlier writes submitted by the same client.

**Review by ogzhanolguncu — APPROVED** [review]

**Review by CahidArda — APPROVED** [review]

**Review by CahidArda — APPROVED** [review]

**Review by ogzhanolguncu — APPROVED** [review]
